File: langs/nodejs/forgehx/src/engine.js
// ForgeHash-X v0 memory-hard engine (SPECIFICATION_X §§6–12).

import { rotl, permute } from "./forgeperm.js";
import { ForgeX } from "./forgex.js";
import {
  BLOCK_SIZE,
  WORDS_PER_BLOCK,
  MIN_SALT_LENGTH,
  MAX_SALT_LENGTH,
} from "./params.js";

const TAG_SEED = "ForgeX/v0/seed";
const TAG_EXPAND = "ForgeX/v0/expand";
const TAG_FINAL = "ForgeX/v0/final";
const TAG_OUTPUT = "ForgeX/v0/output";

const MASK64 = 0xffffffffffffffffn;

function le32(v) {
  const b = Buffer.alloc(4);
  b.writeUInt32LE(v >>> 0, 0);
  return b;
}

function le64(v) {
  const b = Buffer.alloc(8);
  b.writeBigUInt64LE(BigInt(v) & MASK64, 0);
  return b;
}

/** FastRange(x, n) = High64(x * n) using a 128-bit intermediate product. */
export function fastRange(x, n) {
  return ((x & MASK64) * (n & MASK64)) >> 64n;
}

function validateSalt(salt) {
  if (salt.length < MIN_SALT_LENGTH || salt.length > MAX_SALT_LENGTH) {
    throw new RangeError("ForgeHash-X: salt length out of range");
  }
}

function buildMaterial(password, salt, params) {
  return Buffer.concat([
    le32(0),
    le32(params.memoryKiB),
    le32(params.iterations),
    le32(params.parallelism),
    le32(params.outputLength),
    le64(password.length),
    password,
    le32(salt.length),
    salt,
  ]);
}

export function deriveSeed(password, salt, params) {
  params.validate();
  validateSalt(salt);
  return ForgeX.hash(TAG_SEED, buildMaterial(password, salt, params));
}

function blockOffset(lane, index, blocksPerLane) {
  return (lane * blocksPerLane + index) * WORDS_PER_BLOCK;
}

function getBlock(memory, lane, index, blocksPerLane) {
  const start = blockOffset(lane, index, blocksPerLane);
  return memory.slice(start, start + WORDS_PER_BLOCK);
}

function setBlock(memory, lane, index, blocksPerLane, words) {
  const start = blockOffset(lane, index, blocksPerLane);
  for (let i = 0; i < WORDS_PER_BLOCK; i++) {
    memory[start + i] = words[i];
  }
}

function bytesToWords(data) {
  const words = new Array(data.length / 8);
  for (let i = 0; i < words.length; i++) {
    words[i] = data.readBigUInt64LE(i * 8);
  }
  return words;
}

function wordsToBytes(words) {
  const buf = Buffer.alloc(words.length * 8);
  for (let i = 0; i < words.length; i++) {
    buf.writeBigUInt64LE(words[i] & MASK64, i * 8);
  }
  return buf;
}

function selectReference(
  previous,
  pass,
  slice,
  currentLane,
  blockIndex,
  parallelism,
  blocksPerLane,
  sliceLength
) {
  const addressWord =
    (previous[0] ^
      rotl(previous[9], 19) ^
      previous[31] ^
      BigInt(pass >>> 0) ^
      rotl(BigInt(blockIndex >>> 0), 11)) &
    MASK64;

  let lane;
  if (parallelism === 1) {
    lane = 0;
  } else if (blockIndex % 16 === 0) {
    lane = Number(fastRange(previous[1], BigInt(parallelism)));
  } else {
    lane = currentLane;
  }

  function allowed(refLane) {
    if (refLane === currentLane) {
      return pass === 0 ? blockIndex : blocksPerLane;
    }
    return slice * sliceLength;
  }

  let a = allowed(lane);
  if (a === 0) {
    lane = currentLane;
    a = allowed(lane);
  }
  return [lane, Number(fastRange(addressWord, BigInt(a)))];
}

function blockMix(inp, pass, lane, blockIndex) {
  const out = new Array(WORDS_PER_BLOCK);
  for (let k = 0; k < 4; k++) {
    const chunk = inp.slice(k * 16, (k + 1) * 16);
    const state = chunk.slice();
    state[0] ^= BigInt(pass) & MASK64;
    state[1] ^= BigInt(lane) & MASK64;
    state[2] ^= BigInt(blockIndex) & MASK64;
    state[3] ^= rotl(BigInt(pass + blockIndex + k) & MASK64, 13);
    permute(state);
    for (let i = 0; i < 16; i++) {
      out[k * 16 + i] = (state[i] ^ chunk[i]) & MASK64;
    }
  }
  return out;
}

function processLaneSlice(
  memory,
  pass,
  slice,
  lane,
  parallelism,
  blocksPerLane,
  sliceLength
) {
  let start = slice * sliceLength;
  const end = start + sliceLength;
  if (pass === 0 && slice === 0) {
    start = 2;
  }
  for (let blockIndex = start; blockIndex < end; blockIndex++) {
    const previousIndex = blockIndex > 0 ? blockIndex - 1 : blocksPerLane - 1;
    const prev = getBlock(memory, lane, previousIndex, blocksPerLane);
    const [refLane, refIndex] = selectReference(
      prev,
      pass,
      slice,
      lane,
      blockIndex,
      parallelism,
      blocksPerLane,
      sliceLength
    );
    const reference = getBlock(memory, refLane, refIndex, blocksPerLane);
    const combined = new Array(WORDS_PER_BLOCK);
    for (let w = 0; w < WORDS_PER_BLOCK; w++) {
      combined[w] = (prev[w] ^ reference[w]) & MASK64;
    }
    const mixed = blockMix(combined, pass, lane, blockIndex);
    if (pass === 0) {
      setBlock(memory, lane, blockIndex, blocksPerLane, mixed);
    } else {
      const cur = getBlock(memory, lane, blockIndex, blocksPerLane);
      const xored = new Array(WORDS_PER_BLOCK);
      for (let w = 0; w < WORDS_PER_BLOCK; w++) {
        xored[w] = (cur[w] ^ mixed[w]) & MASK64;
      }
      setBlock(memory, lane, blockIndex, blocksPerLane, xored);
    }
  }
}

export function deriveHash(password, salt, params) {
  params.validate();
  validateSalt(salt);

  const seed = ForgeX.hash(TAG_SEED, buildMaterial(password, salt, params));
  const parallelism = params.parallelism;
  const blocksPerLane = params.blocksPerLane;
  const sliceLength = params.sliceLength;
  const memory = new Array(params.blockCount * WORDS_PER_BLOCK).fill(0n);

  for (let lane = 0; lane < parallelism; lane++) {
    for (let i = 0; i < 2; i++) {
      const expandInput = Buffer.concat([seed, le32(lane), le32(i)]);
      const blockBytes = ForgeX.xof(TAG_EXPAND, expandInput, BLOCK_SIZE);
      setBlock(memory, lane, i, blocksPerLane, bytesToWords(blockBytes));
    }
  }

  for (let pass = 0; pass < params.iterations; pass++) {
    for (let slice = 0; slice < 4; slice++) {
      for (let lane = 0; lane < parallelism; lane++) {
        processLaneSlice(
          memory,
          pass,
          slice,
          lane,
          parallelism,
          blocksPerLane,
          sliceLength
        );
      }
    }
  }

  const fold = new Array(WORDS_PER_BLOCK).fill(0n);
  const last = blocksPerLane - 1;
  const q1 = Math.floor(blocksPerLane / 4);
  const q2 = Math.floor(blocksPerLane / 2);
  const q3 = Math.floor((blocksPerLane * 3) / 4);
  for (let lane = 0; lane < parallelism; lane++) {
    for (const index of [last, q1, q2, q3]) {
      const blk = getBlock(memory, lane, index, blocksPerLane);
      for (let w = 0; w < WORDS_PER_BLOCK; w++) {
        fold[w] ^= blk[w];
      }
    }
  }

  const foldBytes = wordsToBytes(fold);
  const finalInput = Buffer.concat([
    seed,
    foldBytes,
    le32(params.memoryKiB),
    le32(params.iterations),
    le32(params.parallelism),
    le32(params.outputLength),
  ]);
  const root = ForgeX.hash(TAG_FINAL, finalInput);
  return ForgeX.xof(TAG_OUTPUT, Buffer.concat([root, seed]), params.outputLength);
}

File: langs/rust/forgehx/src/engine.rs
//! ForgeHash-X v0 memory-hard engine (SPECIFICATION_X §§6–12).

use crate::error::Error;
use crate::forgeperm::{self, WORDS as PERM_WORDS};
use crate::forgex::ForgeX;
use crate::params::{Params, BLOCK_SIZE, MAX_SALT_LENGTH, MIN_SALT_LENGTH, WORDS_PER_BLOCK};

const TAG_SEED: &str = "ForgeX/v0/seed";
const TAG_EXPAND: &str = "ForgeX/v0/expand";
const TAG_FINAL: &str = "ForgeX/v0/final";
const TAG_OUTPUT: &str = "ForgeX/v0/output";

#[inline]
fn fast_range(x: u64, n: u64) -> u64 {
    ((x as u128) * (n as u128) >> 64) as u64
}

fn validate_salt(salt: &[u8]) -> Result<(), Error> {
    if !(MIN_SALT_LENGTH..=MAX_SALT_LENGTH).contains(&salt.len()) {
        return Err(Error::Params("salt length out of range"));
    }
    Ok(())
}

fn build_material(password: &[u8], salt: &[u8], params: &Params) -> Vec<u8> {
    let mut buf = Vec::with_capacity(4 * 5 + 8 + password.len() + 4 + salt.len());
    buf.extend_from_slice(&0u32.to_le_bytes()); // algorithm family id for X sandbox
    buf.extend_from_slice(&(params.memory_kib as u32).to_le_bytes());
    buf.extend_from_slice(&(params.iterations as u32).to_le_bytes());
    buf.extend_from_slice(&(params.parallelism as u32).to_le_bytes());
    buf.extend_from_slice(&(params.output_length as u32).to_le_bytes());
    buf.extend_from_slice(&(password.len() as u64).to_le_bytes());
    buf.extend_from_slice(password);
    buf.extend_from_slice(&(salt.len() as u32).to_le_bytes());
    buf.extend_from_slice(salt);
    buf
}

pub fn derive_seed(password: &[u8], salt: &[u8], params: &Params) -> Result<[u8; 32], Error> {
    params.validate()?;
    validate_salt(salt)?;
    Ok(ForgeX::hash(TAG_SEED, &build_material(password, salt, params)))
}

fn bytes_to_words(data: &[u8]) -> Vec<u64> {
    data.chunks_exact(8)
        .map(|c| u64::from_le_bytes(c.try_into().unwrap()))
        .collect()
}

fn words_to_bytes(words: &[u64]) -> Vec<u8> {
    let mut out = Vec::with_capacity(words.len() * 8);
    for w in words {
        out.extend_from_slice(&w.to_le_bytes());
    }
    out
}

struct Memory {
    words: Vec<u64>,
    blocks_per_lane: usize,
}

impl Memory {
    fn new(block_count: usize, blocks_per_lane: usize) -> Self {
        Self {
            words: vec![0u64; block_count * WORDS_PER_BLOCK],
            blocks_per_lane,
        }
    }

    fn block(&self, lane: usize, index: usize) -> &[u64] {
        let start = (lane * self.blocks_per_lane + index) * WORDS_PER_BLOCK;
        &self.words[start..start + WORDS_PER_BLOCK]
    }

    fn block_mut(&mut self, lane: usize, index: usize) -> &mut [u64] {
        let start = (lane * self.blocks_per_lane + index) * WORDS_PER_BLOCK;
        &mut self.words[start..start + WORDS_PER_BLOCK]
    }

    fn set_block(&mut self, lane: usize, index: usize, words: &[u64]) {
        self.block_mut(lane, index).copy_from_slice(words);
    }
}

fn select_reference(
    previous: &[u64],
    pass: usize,
    slice: usize,
    current_lane: usize,
    block_index: usize,
    parallelism: usize,
    blocks_per_lane: usize,
    slice_length: usize,
) -> (usize, usize) {
    let address_word = previous[0]
        ^ forgeperm::rotl(previous[9], 19)
        ^ previous[31]
        ^ (pass as u64 & 0xFFFF_FFFF)
        ^ forgeperm::rotl(block_index as u64 & 0xFFFF_FFFF, 11);

    let mut lane = if parallelism == 1 {
        0
    } else if block_index % 16 == 0 {
        fast_range(previous[1], parallelism as u64) as usize
    } else {
        current_lane
    };

    let allowed = |ref_lane: usize| -> u64 {
        if ref_lane == current_lane {
            if pass == 0 {
                block_index as u64
            } else {
                blocks_per_lane as u64
            }
        } else {
            (slice * slice_length) as u64
        }
    };

    let mut a = allowed(lane);
    if a == 0 {
        lane = current_lane;
        a = allowed(lane);
    }
    let ref_index = fast_range(address_word, a) as usize;
    (lane, ref_index)
}

fn block_mix(inp: &[u64], pass: usize, lane: usize, block_index: usize) -> [u64; WORDS_PER_BLOCK] {
    let mut out = [0u64; WORDS_PER_BLOCK];
    for k in 0..4 {
        let chunk_start = k * PERM_WORDS;
        let mut state = [0u64; PERM_WORDS];
        state.copy_from_slice(&inp[chunk_start..chunk_start + PERM_WORDS]);
        let chunk = state;
        state[0] ^= pass as u64;
        state[1] ^= lane as u64;
        state[2] ^= block_index as u64;
        state[3] ^= forgeperm::rotl(
            (pass as u64)
                .wrapping_add(block_index as u64)
                .wrapping_add(k as u64),
            13,
        );
        forgeperm::permute(&mut state);
        for i in 0..PERM_WORDS {
            out[chunk_start + i] = state[i] ^ chunk[i];
        }
    }
    out
}

fn process_lane_slice(
    memory: &mut Memory,
    pass: usize,
    slice: usize,
    lane: usize,
    parallelism: usize,
    blocks_per_lane: usize,
    slice_length: usize,
) {
    let mut start = slice * slice_length;
    let end = start + slice_length;
    if pass == 0 && slice == 0 {
        start = 2;
    }
    for block_index in start..end {
        let previous_index = if block_index > 0 {
            block_index - 1
        } else {
            blocks_per_lane - 1
        };
        let mut prev = [0u64; WORDS_PER_BLOCK];
        prev.copy_from_slice(memory.block(lane, previous_index));
        let (ref_lane, ref_index) = select_reference(
            &prev,
            pass,
            slice,
            lane,
            block_index,
            parallelism,
            blocks_per_lane,
            slice_length,
        );
        let mut combined = [0u64; WORDS_PER_BLOCK];
        let reference = memory.block(ref_lane, ref_index);
        for w in 0..WORDS_PER_BLOCK {
            combined[w] = prev[w] ^ reference[w];
        }
        let mixed = block_mix(&combined, pass, lane, block_index);
        if pass == 0 {
            memory.set_block(lane, block_index, &mixed);
        } else {
            let cur = memory.block_mut(lane, block_index);
            for w in 0..WORDS_PER_BLOCK {
                cur[w] ^= mixed[w];
            }
        }
    }
}

pub fn derive_hash(password: &[u8], salt: &[u8], params: &Params) -> Result<Vec<u8>, Error> {
    params.validate()?;
    validate_salt(salt)?;

    let seed = ForgeX::hash(TAG_SEED, &build_material(password, salt, params));
    let parallelism = params.parallelism;
    let blocks_per_lane = params.blocks_per_lane();
    let slice_length = params.slice_length();
    let mut memory = Memory::new(params.block_count(), blocks_per_lane);

    for lane in 0..parallelism {
        for i in 0..2 {
            let mut expand_input = Vec::with_capacity(40);
            expand_input.extend_from_slice(&seed);
            expand_input.extend_from_slice(&(lane as u32).to_le_bytes());
            expand_input.extend_from_slice(&(i as u32).to_le_bytes());
            let block_bytes = ForgeX::xof(TAG_EXPAND, &expand_input, BLOCK_SIZE);
            memory.set_block(lane, i, &bytes_to_words(&block_bytes));
        }
    }

    for pass in 0..params.iterations {
        for slice in 0..4 {
            for lane in 0..parallelism {
                process_lane_slice(
                    &mut memory,
                    pass,
                    slice,
                    lane,
                    parallelism,
                    blocks_per_lane,
                    slice_length,
                );
            }
        }
    }

    let mut fold = [0u64; WORDS_PER_BLOCK];
    let last = blocks_per_lane - 1;
    let q1 = blocks_per_lane / 4;
    let q2 = blocks_per_lane / 2;
    let q3 = (blocks_per_lane * 3) / 4;
    for lane in 0..parallelism {
        for &index in &[last, q1, q2, q3] {
            let blk = memory.block(lane, index);
            for w in 0..WORDS_PER_BLOCK {
                fold[w] ^= blk[w];
            }
        }
    }

    let fold_bytes = words_to_bytes(&fold);
    let mut final_input = Vec::with_capacity(32 + 512 + 16);
    final_input.extend_from_slice(&seed);
    final_input.extend_from_slice(&fold_bytes);
    final_input.extend_from_slice(&(params.memory_kib as u32).to_le_bytes());
    final_input.extend_from_slice(&(params.iterations as u32).to_le_bytes());
    final_input.extend_from_slice(&(params.parallelism as u32).to_le_bytes());
    final_input.extend_from_slice(&(params.output_length as u32).to_le_bytes());

    let root = ForgeX::hash(TAG_FINAL, &final_input);
    let mut output_input = Vec::with_capacity(64);
    output_input.extend_from_slice(&root);
    output_input.extend_from_slice(&seed);
    Ok(ForgeX::xof(TAG_OUTPUT, &output_input, params.output_length))
}
